ethmonitor/ethmonitor.go

```diff
@@ -684,6 +684,16 @@ func (m *Monitor) addLogs(ctx context.Context, blocks Blocks) {
 				block.LogsPayload = m.setPayload(logsPayload)
 				block.OK = true
 				continue
+			} else if len(logs) == 0 && block.Bloom() != (types.Bloom{}) {
+				if m.cache != nil {
+					// delete the cache entry for these block logs as we don't want to store '[]'
+					// when we actually expect logs, as the block bloom filter tells us we should.
+					key := cacheKeyBlockLogs(m.chainID, blockHash, topics)
+					err := m.cache.Delete(ctx, key)
+					if err != nil {
+						m.log.Warnf("ethmonitor: error deleting block logs cache for block %s due to: '%v'", blockHash.String(), err)
+					}
```

Comment:
Nice catch. Looks good.

One small suggestion would be to pass a boolean into `m.filterLogs()` indicating that we expect some logs, so the SET and DELETE logic lives closer to each other.

Or can we even avoid setting the empty `[]` cache in the first place?

```diff
-logs, logsPayload, err := m.filterLogs(tctx, blockHash, topics)
 
+expectLogs := block.Bloom() != (types.Bloom{}) 
+logs, logsPayload, err := m.filterLogs(tctx, blockHash, topics, expectLogs)
```

Reply:
unfortunately not, the way this works internally there are different subsystems even though it does't appear so. its best to keep it as I know it works

Reply:
thanks for the suggestion, I updated the PR to incorporate your idea :) turns out it fit well